File: sdk/typescript/src/providers/migrations.ts
import {
  AlreadyExistsError,
  type ColumnSchema,
  type IndexedDB,
  type IndexSchema,
  type Key,
  type KeyRange,
  NotFoundError,
  type Record as DBRecord,
} from "./indexeddb.ts";

const DEFAULT_LEDGER_STORE = "_gestalt_migrations";
const LEDGER_KEY_COLUMN = "revision_id";
const LEDGER_APPLIED_COLUMN = "applied_at";

export type Query = Key | KeyRange;

/**
 * Declarative schema for one store the SDK ensures exists and matches. Columns
 * are fixed at creation (typically `{ id, payload }`); record-shape evolution
 * lives inside the payload and needs no migration.
 */
export interface StoreDeclaration {
  name: string;
  columns?: ColumnSchema[];
  /** Indexes created together with the store. */
  indexes?: IndexSchema[];
}

export interface AddIndexDeclaration {
  store: string;
  index: IndexSchema;
}

export interface IndexRef {
  store: string;
  name: string;
}

/**
 * A declarative schema revision. Applying it converges the database toward the
 * declared state using only the four closed operations — add/drop a store,
 * add/remove an index — so applying it once or many times yields the same
 * result. Drops must be explicit; they are never inferred from omission.
 */
export interface SchemaDeclaration {
  /** Stores to create if absent (with their initial indexes). */
  stores?: StoreDeclaration[];
  /** Indexes to add to already-existing stores. */
  addIndexes?: AddIndexDeclaration[];
  /** Stores to drop if present. Destructive and irreversible. */
  dropStores?: string[];
  /** Indexes to remove if present. */
  dropIndexes?: IndexRef[];
}

/**
 * A declarative schema revision. The SDK diffs desired-vs-current and applies
 * only the delta.
 */
export interface SchemaRevision {
  id: string;
  schema: SchemaDeclaration;
  backfill?: never;
}

export interface BackfillTransform {
  from: string;
  into: string;
  value: (row: DBRecord) => DBRecord;
}

export interface BackfillRevision {
  id: string;
  backfill: BackfillTransform;
  schema?: never;
}

/** One author-declared migration, identified by its stable, ordered `id`. */
export type Revision = SchemaRevision | BackfillRevision;

export interface MigrationRunOptions {
  revisions: Revision[];
  /** Name of the IndexedDB binding to migrate. */
  dbBinding?: string;
  /**
   * Ledger store name. When omitted, {@link runMigrations} defaults to
   * `_gestalt_migrations`; provider startup derives a per-provider store from
   * the configured provider name.
   */
  ledgerStore?: string;
}

/**
 * Derive the default migration ledger store for a configured provider name.
 */
export function providerMigrationLedgerStore(providerName: string): string {
  let normalized = providerName.trim();
  const slash = normalized.lastIndexOf("/");
  if (slash >= 0) {
    normalized = normalized.slice(slash + 1);
  }
  if (!normalized) {
    return DEFAULT_LEDGER_STORE;
  }
  const slug = migrationLedgerSlug(normalized);
  if (!slug) {
    return DEFAULT_LEDGER_STORE;
  }
  const snake = slug
    .replace(/([a-z0-9])([A-Z])/g, "$1_$2")
    .replace(/-/g, "_")
    .toLowerCase();
  return `${snake}_migrations`;
}

function migrationLedgerSlug(value: string): string {
  let slug = "";
  for (const char of value) {
    if (
      (char >= "A" && char <= "Z") ||
      (char >= "a" && char <= "z") ||
      (char >= "0" && char <= "9") ||
      char === "." ||
      char === "_" ||
      char === "-"
    ) {
      slug += char;
    } else {
      slug += "-";
    }
  }
  return slug.replace(/^-+|-+$/g, "");
}

/** Resolve the IndexedDB binding for migration runs. */
export function resolveMigrationDbBinding(
  options: MigrationRunOptions,
  config: Record<string, unknown>,
): string {
  const explicit = String(options.dbBinding ?? "").trim();
  if (explicit) {
    return explicit;
  }
  const configBinding = config.indexeddb;
  return typeof configBinding === "string" ? configBinding.trim() : "";
}

/**
 * Outcome of a migration run: the revision ids applied this run, in order, and
 * the declared head afterwards (empty when no revisions are declared).
 */
export interface MigrationResult {
  applied: string[];
  head: string;
}

/**
 * Raised when migrations cannot proceed. Carries the triage fields the process
 * surfaces on failure: the current (applied head) and attempted (declared head)
 * revision.
 */
export class MigrationError extends Error {
  readonly current: string | undefined;
  readonly attempted: string | undefined;
  readonly cause: unknown;

  constructor(
    message: string,
    options?: {
      current?: string | undefined;
      attempted?: string | undefined;
      cause?: unknown;
    },
  ) {
    super(message);
    this.name = "MigrationError";
    this.current = options?.current;
    this.attempted = options?.attempted;
    this.cause = options?.cause;
  }
}

/**
 * Brings the database up to head. Concurrent instances may run this at once;
 * that is safe because every revision is idempotent by construction, so
 * concurrent runs converge to the same state and the ledger records each
 * revision once. Idempotent — a no-op when nothing is pending. Returns the
 * revision ids applied this run and the declared head.
 */
export async function runMigrations(
  db: IndexedDB,
  options: MigrationRunOptions,
): Promise<MigrationResult> {
  const revisions = validateRevisions(options.revisions);
  if (revisions.length === 0) {
    return { applied: [], head: "" };
  }

  const ledgerStore = options.ledgerStore?.trim() || DEFAULT_LEDGER_STORE;
  await ensureLedgerStore(db, ledgerStore);

  const applied = new Set(await db.objectStore(ledgerStore).getAllKeys());
  assertNotAheadOfCode(revisions, applied);
  assertContiguousPrefix(revisions, applied);

  const attemptedHead = revisions[revisions.length - 1]?.id ?? "";
  let current = latestAppliedId(revisions, applied);
  const appliedNow: string[] = [];
  for (const revision of revisions) {
    if (applied.has(revision.id)) {
      continue;
    }
    try {
      await applyRevision(db, revision);
      await recordRevision(db, ledgerStore, revision.id);
      appliedNow.push(revision.id);
      current = revision.id;
    } catch (error) {
      if (error instanceof MigrationError) {
        throw error;
      }
      throw new MigrationError(
        `migration ${JSON.stringify(revision.id)} failed: ${errorText(error)}`,
        { current, attempted: attemptedHead, cause: error },
      );
    }
  }
  return { applied: appliedNow, head: attemptedHead };
}

function validateRevisions(revisions: Revision[]): Revision[] {
  const seen = new Set<string>();
  const normalized: Revision[] = [];
  for (const revision of revisions) {
    const id = revision.id?.trim();
    if (!id) {
      throw new MigrationError("every revision needs a non-empty id");
    }
    if (seen.has(id)) {
      throw new MigrationError(`duplicate revision id ${JSON.stringify(id)}`);
    }
    seen.add(id);
    const hasSchema = "schema" in revision && revision.schema != null;
    const backfill = revision.backfill;
    const hasBackfill = backfill != null;
    if (hasSchema === hasBackfill) {
      throw new MigrationError(
        `revision ${JSON.stringify(id)} must declare exactly one of "schema" or "backfill"`,
      );
    }
    if (backfill && backfill.from === backfill.into) {
      throw new MigrationError(
        `revision ${JSON.stringify(id)} backfill "from" and "into" must differ: a ` +
          `backfill reads an immutable source and writes a distinct target, so it ` +
          `cannot read its own output and is idempotent by construction`,
      );
    }
    normalized.push({ ...revision, id });
  }
  return normalized;
}

function revisionNamespaces(revisions: Revision[]): {
  prefixes: Set<string>;
  hasFlat: boolean;
} {
  const prefixes = new Set<string>();
  let hasFlat = false;
  for (const revision of revisions) {
    const id = revision.id.trim();
    if (!id) {
      continue;
    }
    const slash = id.lastIndexOf("/");
    if (slash >= 0) {
      prefixes.add(id.slice(0, slash + 1));
      continue;
    }
    hasFlat = true;
  }
  return { prefixes, hasFlat };
}

function ledgerIDDirectoryPrefix(id: string): string {
  const slash = id.lastIndexOf("/");
  if (slash < 0) {
    return "";
  }
  return id.slice(0, slash + 1);
}

function ledgerIDOwnedByProvider(
  id: string,
  prefixes: Set<string>,
  hasFlat: boolean,
): boolean {
  if (id.includes("/")) {
    return prefixes.has(ledgerIDDirectoryPrefix(id));
  }
  return hasFlat && prefixes.size === 0;
}

function assertNotAheadOfCode(
  revisions: Revision[],
  applied: Set<string>,
): void {
  const declared = new Set(revisions.map((revision) => revision.id));
  const { prefixes, hasFlat } = revisionNamespaces(revisions);
  const unknown = [...applied]
    .filter(
      (id) =>
        !declared.has(id) && ledgerIDOwnedByProvider(id, prefixes, hasFlat),
    )
    .sort();
  if (unknown.length > 0) {
    const attemptedHead = revisions[revisions.length - 1]?.id;
    throw new MigrationError(
      `ledger is ahead of code: applied revision(s) ${unknown
        .map((id) => JSON.stringify(id))
        .join(", ")} are not declared by this binary. Roll forward to a binary ` +
        `that defines them, or manually undo them and delete their ledger rows.`,
      { current: unknown[unknown.length - 1], attempted: attemptedHead },
    );
  }
}

function assertContiguousPrefix(
  revisions: Revision[],
  applied: Set<string>,
): void {
  let sawUnapplied = false;
  const outOfOrder: string[] = [];
  for (const revision of revisions) {
    if (applied.has(revision.id)) {
      if (sawUnapplied) {
        outOfOrder.push(revision.id);
      }
    } else {
      sawUnapplied = true;
    }
  }
  if (outOfOrder.length > 0) {
    throw new MigrationError(
      `ledger has gaps: revision(s) ${outOfOrder
        .map((id) => JSON.stringify(id))
        .join(", ")} are applied but an earlier declared revision is not. ` +
        `Revisions are an append-only ledger — a new revision must be added after ` +
        `all applied ones, never inserted or reordered before them.`,
      { attempted: revisions[revisions.length - 1]?.id ?? "" },
    );
  }
}

function latestAppliedId(
  revisions: Revision[],
  applied: Set<string>,
): string | undefined {
  let current: string | undefined;
  for (const revision of revisions) {
    if (applied.has(revision.id)) {
      current = revision.id;
    }
  }
  return current;
}

async function ensureLedgerStore(
  db: IndexedDB,
  ledgerStore: string,
): Promise<void> {
  await createStoreIfAbsent(db, {
    name: ledgerStore,
    columns: [
      { name: LEDGER_KEY_COLUMN, primaryKey: true, notNull: true },
      { name: LEDGER_APPLIED_COLUMN, notNull: true },
    ],
  });
}

async function applyRevision(db: IndexedDB, revision: Revision): Promise<void> {
  if (isSchemaRevision(revision)) {
    await applySchema(db, revision.schema);
    return;
  }
  await applyBackfill(db, revision.backfill);
}

async function applyBackfill(
  db: IndexedDB,
  transform: BackfillTransform,
): Promise<void> {
  const target = db.objectStore(transform.into);
  const cursor = await db.objectStore(transform.from).openCursor();
  if (cursor === null) {
    return;
  }
  try {
    while (await cursor.continue()) {
      const row = cursor.value;
      if (row === undefined) {
        continue;
      }
      await target.put(transform.value(row));
    }
  } finally {
    cursor.close();
  }
}

async function applySchema(
  db: IndexedDB,
  schema: SchemaDeclaration,
): Promise<void> {
  for (const store of schema.stores ?? []) {
    await createStoreIfAbsent(db, store);
  }
  for (const entry of schema.addIndexes ?? []) {
    await createIndexIfAbsent(db, entry.store, entry.index);
  }
  for (const entry of schema.dropIndexes ?? []) {
    await dropIndexIfPresent(db, entry.store, entry.name);
  }
  for (const name of schema.dropStores ?? []) {
    await dropStoreIfPresent(db, name);
  }
}

async function createStoreIfAbsent(
  db: IndexedDB,
  store: StoreDeclaration,
): Promise<void> {
  try {
    await db.createObjectStore(store.name, {
      ...(store.columns !== undefined ? { columns: store.columns } : {}),
      ...(store.indexes !== undefined ? { indexes: store.indexes } : {}),
    });
  } catch (error) {
    if (!(error instanceof AlreadyExistsError)) {
      throw error;
    }
  }
}

async function createIndexIfAbsent(
  db: IndexedDB,
  store: string,
  index: IndexSchema,
): Promise<void> {
  try {
    await db.createIndex(store, index);
  } catch (error) {
    if (!(error instanceof AlreadyExistsError)) {
      throw error;
    }
  }
}

async function dropStoreIfPresent(db: IndexedDB, name: string): Promise<void> {
  try {
    await db.deleteObjectStore(name);
  } catch (error) {
    if (!(error instanceof NotFoundError)) {
      throw error;
    }
  }
}

async function dropIndexIfPresent(
  db: IndexedDB,
  store: string,
  name: string,
): Promise<void> {
  try {
    await db.deleteIndex(store, name);
  } catch (error) {
    if (!(error instanceof NotFoundError)) {
      throw error;
    }
  }
}

async function recordRevision(
  db: IndexedDB,
  ledgerStore: string,
  id: string,
): Promise<void> {
  await db.objectStore(ledgerStore).put({
    [LEDGER_KEY_COLUMN]: id,
    [LEDGER_APPLIED_COLUMN]: new Date().toISOString(),
  });
}

function isSchemaRevision(revision: Revision): revision is SchemaRevision {
  return "schema" in revision && revision.schema != null;
}

function errorText(error: unknown): string {
  if (error instanceof Error) {
    return error.message;
  }
  return String(error);
}

File: sdk/go/migrations/migrations.go
package migrations

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"sort"
	"strings"
	"time"

	"github.com/valon-technologies/gestalt/sdk/go/indexeddb"
)

// Run brings the database up to the declared head revision.
func Run(ctx context.Context, db indexeddb.Database, opts RunOptions) (Result, error) {
	revisions, err := validateRevisions(opts.Revisions)
	if err != nil {
		return Result{}, err
	}
	if len(revisions) == 0 {
		return Result{Applied: nil, Head: ""}, nil
	}

	ledgerStore := strings.TrimSpace(opts.LedgerStore)
	if ledgerStore == "" {
		ledgerStore = defaultLedgerStore
	}
	if err := ensureLedgerStore(ctx, db, ledgerStore); err != nil {
		return Result{}, err
	}

	appliedKeys, err := db.ObjectStore(ledgerStore).GetAllKeys(ctx, nil)
	if err != nil {
		return Result{}, err
	}
	applied := make(map[string]struct{}, len(appliedKeys))
	for _, id := range appliedKeys {
		applied[id] = struct{}{}
	}

	if err := assertNotAheadOfCode(revisions, applied); err != nil {
		return Result{}, err
	}
	if err := assertContiguousPrefix(revisions, applied); err != nil {
		return Result{}, err
	}

	attemptedHead := revisions[len(revisions)-1].ID
	current := latestAppliedID(revisions, applied)
	appliedNow := make([]string, 0, len(revisions))
	for _, revision := range revisions {
		if _, ok := applied[revision.ID]; ok {
			continue
		}
		if err := applyRevision(ctx, db, revision); err != nil {
			var migErr *MigrationError
			if errors.As(err, &migErr) {
				return Result{}, err
			}
			return Result{}, &MigrationError{
				Message:   fmt.Sprintf("migration %s failed: %s", jsonString(revision.ID), errorText(err)),
				Current:   current,
				Attempted: attemptedHead,
				Cause:     err,
			}
		}
		if err := recordRevision(ctx, db, ledgerStore, revision.ID); err != nil {
			var migErr *MigrationError
			if errors.As(err, &migErr) {
				return Result{}, err
			}
			return Result{}, &MigrationError{
				Message:   fmt.Sprintf("migration %s failed: %s", jsonString(revision.ID), errorText(err)),
				Current:   current,
				Attempted: attemptedHead,
				Cause:     err,
			}
		}
		appliedNow = append(appliedNow, revision.ID)
		current = revision.ID
	}
	return Result{Applied: appliedNow, Head: attemptedHead}, nil
}

func validateRevisions(revisions []Revision) ([]Revision, error) {
	seen := make(map[string]struct{}, len(revisions))
	out := make([]Revision, 0, len(revisions))
	for _, revision := range revisions {
		id := strings.TrimSpace(revision.ID)
		if id == "" {
			return nil, &MigrationError{Message: "every revision needs a non-empty id"}
		}
		if _, ok := seen[id]; ok {
			return nil, &MigrationError{Message: fmt.Sprintf("duplicate revision id %s", jsonString(id))}
		}
		seen[id] = struct{}{}

		hasSchema := revision.Schema != nil
		hasBackfill := revision.Backfill != nil
		if hasSchema == hasBackfill {
			return nil, &MigrationError{
				Message: fmt.Sprintf(
					`revision %s must declare exactly one of "schema" or "backfill"`,
					jsonString(id),
				),
			}
		}
		if revision.Backfill != nil && revision.Backfill.From == revision.Backfill.Into {
			return nil, &MigrationError{
				Message: fmt.Sprintf(
					`revision %s backfill "from" and "into" must differ: a backfill reads an immutable source and writes a distinct target, so it cannot read its own output and is idempotent by construction`,
					jsonString(id),
				),
			}
		}
		copied := revision
		copied.ID = id
		out = append(out, copied)
	}
	return out, nil
}

func revisionNamespaces(revisions []Revision) (prefixes map[string]struct{}, hasFlat bool) {
	prefixes = make(map[string]struct{}, len(revisions))
	for _, revision := range revisions {
		id := strings.TrimSpace(revision.ID)
		if id == "" {
			continue
		}
		if i := strings.LastIndex(id, "/"); i >= 0 {
			prefixes[id[:i+1]] = struct{}{}
			continue
		}
		hasFlat = true
	}
	return prefixes, hasFlat
}

func ledgerIDDirectoryPrefix(id string) string {
	if i := strings.LastIndex(id, "/"); i >= 0 {
		return id[:i+1]
	}
	return ""
}

func ledgerIDOwnedByProvider(id string, prefixes map[string]struct{}, hasFlat bool) bool {
	if strings.Contains(id, "/") {
		_, ok := prefixes[ledgerIDDirectoryPrefix(id)]
		return ok
	}
	return hasFlat && len(prefixes) == 0
}

func assertNotAheadOfCode(revisions []Revision, applied map[string]struct{}) error {
	declared := make(map[string]struct{}, len(revisions))
	for _, revision := range revisions {
		declared[revision.ID] = struct{}{}
	}
	prefixes, hasFlat := revisionNamespaces(revisions)
	var unknown []string
	for id := range applied {
		if _, ok := declared[id]; ok {
			continue
		}
		if !ledgerIDOwnedByProvider(id, prefixes, hasFlat) {
			continue
		}
		unknown = append(unknown, id)
	}
	if len(unknown) == 0 {
		return nil
	}
	sort.Strings(unknown)
	attemptedHead := revisions[len(revisions)-1].ID
	return &MigrationError{
		Message: fmt.Sprintf(
			`ledger is ahead of code: applied revision(s) %s are not declared by this binary. Roll forward to a binary that defines them, or manually undo them and delete their ledger rows.`,
			joinJSONStrings(unknown),
		),
		Current:   unknown[len(unknown)-1],
		Attempted: attemptedHead,
	}
}

func assertContiguousPrefix(revisions []Revision, applied map[string]struct{}) error {
	sawUnapplied := false
	var outOfOrder []string
	for _, revision := range revisions {
		if _, ok := applied[revision.ID]; ok {
			if sawUnapplied {
				outOfOrder = append(outOfOrder, revision.ID)
			}
		} else {
			sawUnapplied = true
		}
	}
	if len(outOfOrder) == 0 {
		return nil
	}
	attemptedHead := ""
	if len(revisions) > 0 {
		attemptedHead = revisions[len(revisions)-1].ID
	}
	return &MigrationError{
		Message: fmt.Sprintf(
			`ledger has gaps: revision(s) %s are applied but an earlier declared revision is not. Revisions are an append-only ledger — a new revision must be added after all applied ones, never inserted or reordered before them.`,
			joinJSONStrings(outOfOrder),
		),
		Attempted: attemptedHead,
	}
}

func latestAppliedID(revisions []Revision, applied map[string]struct{}) string {
	var current string
	for _, revision := range revisions {
		if _, ok := applied[revision.ID]; ok {
			current = revision.ID
		}
	}
	return current
}

func ensureLedgerStore(ctx context.Context, db indexeddb.Database, ledgerStore string) error {
	return createStoreIfAbsent(ctx, db, StoreDeclaration{
		Name: ledgerStore,
		Columns: []indexeddb.ColumnDef{
			{Name: ledgerKeyColumn, PrimaryKey: true, NotNull: true},
			{Name: ledgerAppliedCol, NotNull: true},
		},
	})
}

func applyRevision(ctx context.Context, db indexeddb.Database, revision Revision) error {
	if isSchemaRevision(revision) {
		return applySchema(ctx, db, *revision.Schema)
	}
	return applyBackfill(ctx, db, *revision.Backfill)
}

func applyBackfill(ctx context.Context, db indexeddb.Database, transform BackfillTransform) error {
	target := db.ObjectStore(transform.Into)
	cursor, err := db.ObjectStore(transform.From).OpenCursor(ctx, nil, indexeddb.CursorNext)
	if err != nil {
		return err
	}
	defer cursor.Close()

	for cursor.Continue() {
		row, err := cursor.Value()
		if err != nil {
			if errors.Is(err, indexeddb.ErrNotFound) {
				continue
			}
			return err
		}
		if row == nil {
			continue
		}
		if err := target.Put(ctx, transform.Value(row)); err != nil {
			return err
		}
	}
	return cursor.Err()
}

func applySchema(ctx context.Context, db indexeddb.Database, schema SchemaDeclaration) error {
	for _, store := range schema.Stores {
		if err := createStoreIfAbsent(ctx, db, store); err != nil {
			return err
		}
	}
	for _, entry := range schema.AddIndexes {
		if err := createIndexIfAbsent(ctx, db, entry.Store, entry.Index); err != nil {
			return err
		}
	}
	for _, entry := range schema.DropIndexes {
		if err := dropIndexIfPresent(ctx, db, entry.Store, entry.Name); err != nil {
			return err
		}
	}
	for _, name := range schema.DropStores {
		if err := dropStoreIfPresent(ctx, db, name); err != nil {
			return err
		}
	}
	return nil
}

func createStoreIfAbsent(ctx context.Context, db indexeddb.Database, store StoreDeclaration) error {
	opts := indexeddb.ObjectStoreOptions{}
	if store.Columns != nil {
		opts.Columns = store.Columns
	}
	if store.Indexes != nil {
		opts.Indexes = store.Indexes
	}
	_, err := db.CreateObjectStore(ctx, store.Name, opts)
	if err != nil && !errors.Is(err, indexeddb.ErrAlreadyExists) {
		return err
	}
	return nil
}

func createIndexIfAbsent(ctx context.Context, db indexeddb.Database, store string, index indexeddb.IndexSchema) error {
	idxMgr, ok := db.(indexeddb.IndexManager)
	if !ok {
		return fmt.Errorf("indexeddb: create index %q on %q: index manager not supported", index.Name, store)
	}
	err := idxMgr.CreateIndex(ctx, store, indexeddb.IndexDefinition{
		Name:    index.Name,
		KeyPath: index.KeyPath,
		Unique:  index.Unique,
	})
	if err != nil && !errors.Is(err, indexeddb.ErrAlreadyExists) {
		return err
	}
	return nil
}

func dropStoreIfPresent(ctx context.Context, db indexeddb.Database, name string) error {
	err := db.DeleteObjectStore(ctx, name)
	if err != nil && !errors.Is(err, indexeddb.ErrNotFound) {
		return err
	}
	return nil
}

func dropIndexIfPresent(ctx context.Context, db indexeddb.Database, store, name string) error {
	idxMgr, ok := db.(indexeddb.IndexManager)
	if !ok {
		return fmt.Errorf("indexeddb: delete index %q on %q: index manager not supported", name, store)
	}
	err := idxMgr.DeleteIndex(ctx, store, name)
	if err != nil && !errors.Is(err, indexeddb.ErrNotFound) {
		return err
	}
	return nil
}

func recordRevision(ctx context.Context, db indexeddb.Database, ledgerStore, id string) error {
	record := indexeddb.Record{
		ledgerKeyColumn:  id,
		ledgerAppliedCol: time.Now().UTC().Format(time.RFC3339Nano),
	}
	if ledgerKeyColumn != "id" {
		record["id"] = id
	}
	return db.ObjectStore(ledgerStore).Put(ctx, record)
}

func isSchemaRevision(revision Revision) bool {
	return revision.Schema != nil
}

func errorText(err error) string {
	if err == nil {
		return ""
	}
	return err.Error()
}

func jsonString(v string) string {
	b, err := json.Marshal(v)
	if err != nil {
		return fmt.Sprintf("%q", v)
	}
	return string(b)
}

func joinJSONStrings(values []string) string {
	parts := make([]string, len(values))
	for i, v := range values {
		parts[i] = jsonString(v)
	}
	return strings.Join(parts, ", ")
}
